src/status_im/navigation/events.cljs

```diff
@@ -188,3 +188,25 @@
   (let [view-id (if (= view-id :shell-stack) (shell.utils/calculate-view-id) view-id)]
     {:db             (assoc db :view-id view-id)
      :set-view-id-fx view-id}))
+
+(defn navigate-wizard-next-screen
+  [db flow-config current-screen]
+  (first (filter (fn [screen]
+                   (let [skip-step (:skip-step? screen)]
+                     (and (not= (:screen-id screen) current-screen)
+                          (not (when (not (nil? skip-step)) (skip-step db))))))
```

Comment:
Also consider using `some` instead of `first` + `filter`

Reply:
I tend to not like `some` as much as `first` + `filter` because if I forget to explicitly return the value passed to the predicate function, then `some` will only return true or false, but not the element found. With `filter`, I can just use any predicate and not worry about anything.

```clojure
(defn pred
  [n]
  (> n 10))

(some #(when (pred %) %) (range 20)) ; => 11
(some pred (range 20))               ; => true
(first (filter pred (range 20)))     ; => 11
```

`some` is nice sometimes, but I still keep going for `filter` + `first` most of the time.

Probably with thousands of elements to filter I wouldn't use `filter` + `first`, although if the sequence is chunked the performance will be just fine.

Reply:
@ulisesmac ⬆️

Reply:
@ilmotta 

> Probably with thousands of elements to filter I wouldn't use filter + first, although if the sequence is chunked the performance will be just fine.

Well, even if they aren't thousands I think there's no need to process those extra elements.
Yeah, for chunked seqs first + filter is fine, although I've seen most of the data we store in re-frame are vectors.

I do recommend `some` because:

- Code style: IMO looks more straightforward, if we just want the first element, `some` is clearer, a single function instead of two. If we want to filter a whole seq and return a new one then `filter` is fine.

- An easy to apply change.

> if I forget to explicitly return the value passed to the predicate function, then some will only return true or false

I've rarely forgotten to use `some` in the appropiate way, I think they are significanlty different.


We don't have any strict guideline about this, so we can probably just let the developer use their prefered style.

Reply:
@ulisesmac:

> I've rarely forgotten to use some in the appropiate way, I think they are significanlty different.

Yeah, I think I forget about `some` because I don't use it very often. More like it. Will try to consider it more often!

>  I've seen most of the data we store in re-frame are vectors.

When we parse JSON from status-go and convert using `js->clj`, vectors in JSON end up as vectors in Clojure. But the moment we process them, almost always we use lazy functions. For that reason it's unclear to me if vectors are the majority. It's probably the opposite in many parts of the app.

With respect to performance, I was curious to put `some` against `filter` + `first` because most of the time I'm surprised by microbenchmark's results. They defy my intuition.

My take away from the results is that `some` can even be slower even with vectors, but more importantly, both solutions give comparable performance even in much larger vectors than what we find in our app. Hence, it's better to just use whatever the developer prefers and ignore the performance characteristics of both because the differences will be comparable to random noise given our small data sets.

```clojure
(let [pred #(when (> % 10000) %)
      values (into [] (range 20000))]
  (with-progress-reporting
    (bench (some pred values))))
  ;; Execution time mean : 104.531872 µs.
  ;; Execution time std-deviation : 3.226377 µs
  ;; Execution time lower quantile : 97.936634 µs ( 2.5%)
  ;; Execution time upper quantile : 110.225970 µs (97.5%)

(let [pred #(when (> % 10000) %)
      values (range 20000)]
  (with-progress-reporting
    (bench (some pred values))))
  ;; Execution time mean : 84.364265 µs
  ;; Execution time std-deviation : 3.299388 µs
  ;; Execution time lower quantile : 79.224892 µs ( 2.5%)
  ;; Execution time upper quantile : 88.678269 µs (97.5%)

(let [pred #(> % 10000)
      values (into [] (range 20000))]
  (with-progress-reporting
    (bench (first (filter pred values)))))
  ;; Execution time mean : 58.328354 µs.
  ;; Execution time std-deviation : 1.938134 µs
  ;; Execution time lower quantile : 54.784170 µs ( 2.5%)
  ;; Execution time upper quantile : 60.713590 µs (97.5%)

(let [pred #(> % 10000)
      values (range 20000)]
  (with-progress-reporting
    (bench (first (filter pred values)))))
;; Execution time mean : 83.813830 µs.
;; Execution time std-deviation : 2.166592 µs
;; Execution time lower quantile : 79.980113 µs ( 2.5%)
;; Execution time upper quantile : 87.142827 µs (97.5%)
```

Reply:
@ilmotta  
Very interesting results!

Are you measuring this on CLJS? would you mind sharing the lib? :)

> it's better to just use whatever the developer prefers and ignore the performance characteristics of both because the differences will be comparable to random noise given our small data sets.

Totally agree!
Thanks for these comments, I usually share with others the suggestion of `some` instead of `first` + `filter`, mostly because of code style and because some devs don't know `some` exists. Now I'm thinking this suggestion is superflous.

Reply:
Great to know you liked to see the results @ulisesmac!

It's always very tricky to measure anything on cljs. Criterium is the best Clojure library for these microbenchmarks, but it only works on the JVM, and that's what I used because `simple-benchmark` is too flawed in cljs. I have a separate repo I often switch to quickly spin up a REPL and benchmark with criterium.

What I do that I think helps is before comparing cljs & clj functions, I take a look at their implementations. Specifically for `some`, they are identical, so it's a good indicator that results should translate well between both runtimes, but of course that's no guarantee. Looking at the source code is surprisingly humbling, because I get to see I need to study Clojure more.

I have been surprised so many times with the performance of Clojure core functions for good or worse that nowadays I don't trust my intuition anymore.

An important factor in the mobile app is that we usually store and process very small sets of data in the mobile app, often way smaller than 100 items, but we have lots of them. That's very important when we analyze the performance of data structures and the functions we use to operate on them. I usually completely ignore absolute results from criterium because they are bound to be different for cljs and also from real life when running concurrently with other things in the mobile app. But I do look carefully at the relative results. When a function clearly takes half the time to do something, then it catches my attention and then I dig further.

`filter` is more complicated as well because it can return a transducer. But now I appreciate `some` even more for it's implementation simplicity!

Good discussion, thanks for that Ulises.

Reply:
> but it only works on the JVM, 

Oh :cry:  I wouldn't trust too much those benchmarks, besides the Clojure implementation being equal, the internals are too different.

For example, recently I was exploring the unchecked-math in CLJS for performance reasons, but unless something hidden is happening, I think the `unchecked-` operations versions are slower. Weird, I want to think I'm misunderstanding something, but the code is clear:

`unchecked-add` code:
```clojure
(defn ^number unchecked-add
  "Returns the sum of nums. (+) returns 0."
  ([] 0)
  ([x] x)
  ([x y] (cljs.core/unchecked-add x y)) ;;<- calls cljs.core/unchecked-add
  ([x y & more] (reduce unchecked-add (cljs.core/unchecked-add x y) more)))
```

`cljs.core/unchecked-add` code:
```clojure
(core/defmacro ^::ana/numeric unchecked-add
  ([& xs] `(+ ~@xs)))
```
It just calls `+` ._.

```clojure
(core/defmacro ^::ana/numeric +
  ([] 0)
  ([x] (core/list 'js* "(~{})" x))
  ([x y] (core/list 'js* "(~{} + ~{})" x y))
  ([x y & more] `(+ (+ ~x ~y) ~@more)))
```

So, it raised some questions on me:
1. What is the `unchecked-add` difference compared to just `+`?
2. Isn't the nested function calls overhead making this version slower?
3. There's also an overhead in `unchecked-add` when it's called with more than 2 values, because of that `reduce`, it doesn't exists in the regular `+`. why do they do this?
4. If `unchecked-add` just uses `+` at the end, does it mean the regular `+` is unchecked? if so, this is not the same as in JVM Clojure.

The worst part is  `unchecked-add-int` exists but it just calls `+` at the end too.

How many more implementations on CLJS are like this? :cry: 

> I have been surprised so many times with the performance of Clojure core functions for good or worse that nowadays I don't trust my intuition anymore.

Totally share this feeling, it has been the same for me, but I usually test just using `time`, I know it's not too accurate, but it is able to highlight the big differences, good enough for a quick comparison.

Yeah, very interesting discussion :+1:  always a pleasure to discuss Clojure stuff, thanks for it! :smile:

Reply:
Absolutely, math is completely different in these two environments. For math I wouldn't benchmark on the JVM. But for data structure processing functions, like `filter`, `map`, etc the results seem to translate well enough to give us a rough idea. For example, a 100% difference in Criterium probably means something interesting will also happen in the cljs runtime as well.

> Totally share this feeling, it has been the same for me, but I usually test just using time, I know it's not too accurate, but it is able to highlight the big differences, good enough for a quick comparison.

👌🏼 💯